scanpy/tests/test_highly_variable_genes.py

```diff
@@ -54,6 +57,276 @@ def test_highly_variable_genes_basic():
     assert np.all(np.isin(colnames, hvg_df.columns))
 
 
+def _residual_var_reference(adata, clip=None, theta=100):
+    sc.experimental.pp.normalize_pearson_residuals(adata, clip=clip, theta=theta)
+    residuals = adata.X
+    return np.var(residuals, axis=0)
+
+
+@pytest.mark.parametrize(
+    'sparsity_func', [csr_matrix.toarray, csr_matrix], ids=lambda x: x.__name__
+)
+@pytest.mark.parametrize('dtype', ['float32', 'int64'])
+def test_highly_variable_genes_pearson_residuals_inputchecks(sparsity_func, dtype):
+
+    adata = _prepare_pbmc_testdata(sparsity_func, dtype, small=True)
+
+    # depending on check_values, warnings should be raised for non-integer data
+    if dtype == 'float32':
+
+        adata_noninteger = adata.copy()
+        x, y = np.nonzero(adata_noninteger.X)
+        adata_noninteger.X[x[0], y[0]] = 0.5
+
+        # expecting 0 no-int warnings
+        with pytest.warns(None) as record:
+            sc.experimental.pp.highly_variable_genes(
+                adata_noninteger.copy(),
+                flavor='pearson_residuals',
+                n_top_genes=100,
+                check_values=False,
+            )
+        assert len(record) == 0
+
+        # expecting 1 no-int warning
+        with pytest.warns(None) as record:
+            sc.experimental.pp.highly_variable_genes(
+                adata_noninteger.copy(),
+                flavor='pearson_residuals',
+                n_top_genes=100,
+                check_values=True,
+            )
```

Comment:
I think you should be more specific about the warning you are/ are not expecting to see here. Otherwise if pandas starts throwing a warning internally somewhere this test will fail, even though that's not what this test is looking for.

Reply:
added a check for the specific warning to be present/absent